File: scala/nth-prime/src/main/scala/Prime.scala
import collection.mutable
import scala.annotation.tailrec
import scala.collection.mutable.ListBuffer

object Prime {
  type Prime = Int

  def nth(n: Int): Prime =
    loop(n, 3, ListBuffer(2))

  // Sieve of Eratosthenes
  @tailrec
  private def loop(n: Int, candidate: Int, primes: ListBuffer[Int]): Prime = {
    if (primes.size == n) primes.last
    else {
      def notDivisibleBy(i: Int) = candidate % i != 0
      def squareLE(i: Int) = i*i <= candidate

      val relevantPrimes = primes takeWhile squareLE
      val candidateIsPrime = relevantPrimes forall notDivisibleBy
      if (candidateIsPrime) primes += candidate
      val nextCandidate = candidate + 1
      loop(n, nextCandidate, primes)
    }
  }
}

File: haskell/nth-prime/src/Prime.hs
module Prime (nth) where

import Control.Monad (guard)

nth :: Int -> Maybe Integer
nth n = do
  guard  $ n > 0
  return $ nth' n 3 [2]

-- recursive Sieve of Erathosthenes
nth' :: Int -> Integer -> [Integer] -> Integer
nth' n candidate primes
  | n == length primes = last primes
  | otherwise = nth' n nextCandidate nextPrimes
  where
    nextCandidate = succ candidate
    nextPrimes = if candidateIsPrime then primes ++ [candidate] else primes
    candidateIsPrime = all notDivisibleBy relevantPrimes
    notDivisibleBy i = candidate `mod` i > 0
    relevantPrimes = takeWhile squareLE primes
    squareLE i = i*i <= candidate
